sklearn_numba_dpex/common/matmul.py

```diff
@@ -0,0 +1,814 @@
+    - memory layout: the kernel takes `X` and `Y_t` as input and slides a window of
+    items on columns. Taking `X_t` and `Y` as inputs (also `C-contiguous`) and
+    sliding on rows might be better. It seems it can enable prettier memory access
+    patterns both for reading and writing in shared memory. However attempts at
+    this approach have only shown worse performances so far.
+    - This suggests that some assumptions on local memory might be wrong, is local
+    memory C-contiguous ?
+    - coalescing accesses in global memory: it is unclear wether `numba_dpex/SYCL` can
+    effectively coalesce reads of contiguous addresses in global memory, it might not
+    if it's not able to tell at compile time that the input has a contiguous structure
+    in memory. Investigating `SPIR-V` code could confirm/infirm this. If not coalesced,
+    use of functions such as `reinterpret_cast` for `cuda` are required to manually
+    enforce the coalesced reads.
+    - autotuning: maybe better parameters (`work_group_size`, `sub_group_size`,
+    `arithmetic_intensity` multipliers,...) can be found, those parameters might depend
+    on the device. Is it possible to find parameters that reasonably suit all devices ?
+    - Investigating `SPIR-V` code compiled by `numba-dpex` to look for possible
+    issues with the JIT.
+    - Re-ordering work groups (and maybe work items within work groups) to improve
+    cache locality. This as been attempted but have shown no results, while up to a
+    10% improvements is claimed to be possible with this trick.
+
+    See also:
+    - https://siboehm.com/articles/22/CUDA-MMM
+    - https://triton-lang.org/master/getting-started/tutorials/03-matrix-multiplication.html  # noqa
+
+    """
+    # The best values for the parameters defined thereafter depend on the device, but
+    # a stable heuristic is yet to be found.
+    # TODO: find an heuristic for adapting the performance parameters to the device
+    # rather than hardcoding defaults.
+    if work_group_size in [None, "max"]:
+        work_group_size = 128
+    if sub_group_size is None:
+        sub_group_size = 4
+
+    # Arithmetic intensity refers to how much compute a single work item is set to
+    # perform relatively to how much RW operations it needs to execute.
+
+    # The size of the window of results that are computed by a single work item is
+    # scaled by
+    # (arithmetic_intensity_multiplier_X, arithmetic_intensity_multiplier_Y) ratio, at
+    # the cost of loading
+    # `arithmetic_intensity_multiplier_X + arithmetic_intensity_multiplier_Y`
+    # more items in the work group shared memory, per item in this work group.
+
+    # Those values should be set high enough so that the kernel is compute-bound
+    # rather than memory-bound.
+
+    # Setting it too high might pressure local or private memory too much for it to be
+    # worth.
+    arithmetic_intensity_multiplier_X = 4
+    arithmetic_intensity_multiplier_Y = 2
```

Comment:
Let's make those kwargs of `make_matmul_2d_kernel` with the same default values so that it will be possible in the future to do run some kind of auto-tuning grid search.